File: src/app/service/product.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Product } from '../model/product';

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  private baseUrl = 'http://localhost:8080/product/';  

  constructor(private http:HttpClient) { } 
  get4Products(): Observable<Product[]>{
    return this.http.get<Product[]>(`${this.baseUrl}list`);
  }

  get4ProductsDESC(): Observable<Product[]>{
    return this.http.get<Product[]>(`${this.baseUrl}listDESC`);
  }

  getProductsSale(): Observable<Product[]>{
    return this.http.get<Product[]>(`${this.baseUrl}sale`);
  }

   getProductList(pageNo: any, pageSize: any): Observable<Product[]>{
    return this.http.get<Product[]>(`${this.baseUrl}all?pageNo=${pageNo}&pageSize=${pageSize}`);  
  }

  addProduct(product: Product): Observable<Product> {  
    return this.http.post<Product>(`${this.baseUrl}`+'add', product);  
  } 
  saveProduct(formData: FormData): Observable<any> {  
    return this.http.post<Product>(`${this.baseUrl}`+'file', formData);  
  }

  getProductById(id: number): Observable<any> {
    return this.http.get(this.baseUrl + id);
  }

  getProductByCategoryId(categoryId: number): Observable<Product[]>{
    return this.http.get<Product[]>(`${this.baseUrl}getProductByCategory/${categoryId}`);  
  }

  get3ProductByCategoryId(categoryId: number): Observable<Product[]>{
    return this.http.get<Product[]>(`${this.baseUrl}get3ProductByCategory/${categoryId}`);  
  }


  getProductByName(search: string): Observable<any> {
    return this.http.get<Product[]>(`${this.baseUrl}findByName/${search}`);
  }


  updateProduct(id: number, product: Product): Observable<Object> {
    return this.http.put(`${this.baseUrl}update/${id}`, product);
  }

  deleteProduct(id: number): Observable<Object> {
    return this.http.delete(`${this.baseUrl}/delete/${id}`, { responseType: 'text' });
  }

  findByPriceBetween(min: number, max: number): Observable<Product[]>{
    return this.http.get<Product[]>(`${this.baseUrl}${min}/${max}`)
  }
}

File: spring boot backend/src/main/java/com/app/service/ProductService.java
package com.app.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import com.app.model.Product;
import com.app.model.Category;
import com.app.model.Product;
import com.app.repo.ProductRepo;

@Service
public class ProductService {
	@Autowired
	private ProductRepo productRepo;
	
	
	public Product saveProduct(Product product) {
		return productRepo.save(product);
	}
	
	public List<Product> get4Products() {
		return (List<Product>) productRepo.find4Product();
	}
	
	public List<Product> get4ProductsDESC() {
		return (List<Product>) productRepo.find4ProductDESC();
	}
	
	public List<Product> getProductsSale() {
		return (List<Product>) productRepo.findProductSale();
	}
	
	public Page<Product> getProduct(int pageNo, int pageSize){
		Pageable paging = PageRequest.of(pageNo, pageSize);
		return productRepo.findAll(paging);
	}
	
	public Product getProductById(int id) {
		return productRepo.findById(id).orElse(null);
	}
	
	public List<Product> getProductByCategoryId(int categoryId) {
		return productRepo.findProductByCategoryId(categoryId);
	}
	
	public List<Product> get3ProductByCategoryId(int categoryId) {
		return productRepo.find3ProductByCategoryId(categoryId);
	}


	public List<Product> getProductByName(String name) {
		return productRepo.findByName(name);
	}
	

	public String deleteProduct(int id) {
		productRepo.deleteById(id);
		return "Product removed !! " + id;
	}
	
	public Product updateProduct(Product product) {
		Product existingProduct = productRepo.findById(product.getId()).orElse(null);
        existingProduct.setName(product.getName());
        existingProduct.setQuantity(product.getQuantity());
        existingProduct.setPrice(product.getPrice());
        existingProduct.setSale_price(product.getSale_price());
        existingProduct.setDescription(product.getDescription());
        existingProduct.setStatus(product.getStatus());
        existingProduct.setImage(product.getImage());
        existingProduct.setCat(product.getCat());

        return productRepo.save(existingProduct);
    }
	
	
	public List<Product> findByPriceBetween(int min, int max){
		return productRepo.findProductByPrice(min, max);
	}

}
